File: 6-zigzag-conversion.go
func convert(s string, numRows int) string {
    if numRows == 1 {
        return s
    }
    rows := [][]byte{}
    for i := numRows; i > 0; i -= 1 {
        rows = append(rows, []byte{})
    }
    down := true
    line := 0
    for idx := 0; idx < len(s); idx += 1 {
        ch := s[idx]
        rows[line] = append(rows[line], ch)
        if line == 0 {
            down = true
        } else if line == numRows - 1 {
            down = false
        }
        if down {
            line += 1
        } else {
            line -= 1
        }
    }
    res := ""
    for _, row := range rows {
        res += string(row)
    }
    return res
}

File: bwu/string/6-zigzag-conversion.py
class Solution(object):
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        if not s or not numRows:
            return ''
        if numRows == 1:
            return s
            
        n = len(s)
        res = ''
        rows = ['' for _ in xrange(numRows)]
        idx, direction = 0, -1
        for c in s:
            if idx == 0 or idx == numRows - 1:
                direction *= -1
            rows[idx] += c
            idx += 1 * direction
         
        for row in rows:
            res += row
            
        return res
